crates/iota-indexer/src/indexer.rs

```diff
@@ -110,79 +88,103 @@ impl Indexer {
-                    .data_ingestion_path
-                    .clone()
-                    .unwrap_or(tempfile::tempdir().unwrap().keep()),
-                config
-                    .sources
-                    .remote_store_url
-                    .as_ref()
-                    .map(|url| url.as_str().to_owned()),
-                vec![],
-                extra_reader_options,
-            ),
-        );
+        let (mut primary_executor_handle, mut primary_writer_handle) = primary_pipeline
+            .run(
+                data_ingestion_path.clone(),
+                remote_store_url.clone(),
+                extra_reader_options.clone(),
+                cancel.clone(),
+            )
+            .await?;
 
-        tokio::select! {
-            executor_result = &mut executor_handle => {
-                // Executor completed first - cancel snapshot task and check result
-                cancel.cancel();
-                let snapshot_result = tokio::time::timeout(
-                    SHUTDOWN_TIMEOUT,
-                    object_snapshot_task_handle
-                ).await
-                .context("timeout waiting for snapshot task to shutdown");
-                executor_result.context("failed to join data ingestion executor")?.context("data ingestion executor failed")?;
-                snapshot_result?.context("failed to join snapshot task during shutdown")?.context("snapshot task failed during shutdown")?;
+        // Wait for max committable checkpoint > 0 before starting snapshot executor
+        // Also monitor primary_executor_handle - if it finishes, no point in waiting
+        let (mut snapshot_executor_handle, mut snapshot_persist_task_handle) = tokio::select! {
+            snapshot_pipeline_with_snapshottable_data = snapshot_pipeline.wait_for_snapshottable_data(cancel.clone()) => {
+                snapshot_pipeline_with_snapshottable_data?.run(
+                    remote_store_url,
+                    extra_reader_options,
+                    cancel.clone(),
+                ).await?
             },
-            snapshot_result = &mut object_snapshot_task_handle => {
-                // Snapshot task completed first - cancel executor and check result
-                cancel.cancel();
-                let executor_result = tokio::time::timeout(
-                    SHUTDOWN_TIMEOUT,
-                    executor_handle
-                ).await
-                .context("timeout waiting for executor to shutdown");
-                snapshot_result.context("failed to join snapshot task")?.context("snapshot task failed")?;
-                executor_result?.context("failed to join data ingestion executor during shutdown")?.context("data ingestion executor failed during shutdown")?;
+            result = &mut primary_executor_handle => {
+                result.context("failed to join primary executor")?.context("primary executor failed")?;
+                return Ok(());
             }
         };
```

Comment:
This should probably handle the `primary_writer` as well.

But I think that `waiting` for the snapshot pipeline to start could be handled inside the `snapshot::orchestration`. 

I would suggest to reorganize the orchestration a bit, to avoid boilerplate and define different layers of orchestration scope.

* `snapshot::orchestration`: `Pipeline::run` can `wait_for_snapshot_lag` and then handle orchestration between `executor` and `writer` for graceful shutdown. You could for example use an `tokio::select` in `run` after `waiting`. 
* `primary::orchestration`: `Pipeline::run` can handle orchestration between `executor` and `writer` for graceful shutdown. 
* `indexer`: Handle orchestration between `{primary_,snapshot_}pipeline::run` for graceful shutdown.